### ISSUE TITLE
99% CPU utilization on Android when application is stopped

### ISSUE THREAD

**Issue by leonstyhre** [libsdl-org/SDL#8748]
Hi!

I have an issue with the Android port of my application (ES-DE) that I've been unable to figure out. It seems to be caused by an issue in SDL but I'm not entirely sure.

When the application is running everything looks fine and on a specific device I test on it normally consumes around 12-25% of a CPU core. However any time I move away from the application, like when launching an external activity or pressing the home button, CPU utilization jumps to 99% and stays there indefinitely. But as soon as I return to the application and the onStart() and onResume() callbacks are executed CPU utilization drops down to the normal level of 12-25% or so as my native main loop resumes.

The problem seems to occur after the onStop() callback is executed. I have confirmed that the main loop of my native C++ code is indeed stopped so my code can't be generating the CPU load. It really seems like there's a loop somewhere in SDL that consumes all available cycles.

Do you recognize this problem, and is there anything I can test to try to work around the problem?

Edit: This is with SDL 2.28.5 but I've tested with the 3.0.0 prerelease as well and it's the exact same behavior.

Thanks!

**Comment by leonstyhre** [comment]
So after some investigations I found a workaround for this issue but it's very ugly.

The problem seems to be that when SDL pauses the native thread the SDL_SemWait() call in Android_PumpEvents_Blocking() in SDL_androidevents.c somehow indirectly eats all CPU cycles. I first thought it was caused by the sem_wait() call in SDL_SemWait() in SDL_syssem.c but I tried to replace this function with a sem_timedwait() call with a one second wait timer and it still ate all CPU cycles. So I have no idea what is consuming all the cycles.

Anyway, the workaround was to override the onPause() method and make a JNI call that sets a flag to prevent SDL_PollEvent() calls from being executed in my main application loop when the application is being stopped. As long as SDL_PollEvent() is not called the semaphore issue will not get triggered and the process now uses something like 0.3% CPU instead of 99% when stopped.

Overriding onResume() as well and making another JNI call to toggle back the flag makes my main loop resume the SDL_PollEvent() calls and the application therefore works correctly again after being resumed.

Could you please try to find a more proper fix for this? This is a quite tedious workaround and it introduces some unnecessary complexity as I need to check for the presence of the event processing flag on every application cycle. There is also the issue of timing, I'm not sure if my workaround is 100% reliable due to the way that the activity callbacks work in Android.

On the other hand, if I'm doing something wrong with the way I'm handling events on Android and there is a nicer solution to the problem, then please let me know :)
Thanks!

**Comment by slouken** [comment]
@1bsyl, can you take a look?

**Comment by 1bsyl** [comment]
I'll take a look when i have some time.

In the meantimes:
@leonstyhre 
- did you try on another devices?
- do you unset block_on_pause Sdl hint ?
- what command do you use to get the cou info? So that i can try on my side the same way.

The normal case, when apps goes in backgroud is that the code is blocked inside SDL_PollEvents().
Until the app resume/and so the pollevent returns.
Do you use various thread that could still run?

**Comment by leonstyhre** [comment]
Thanks for looking into this!

I did not try it on another device but I could test it on the Android emulator if you want me to, I doubt it will make any difference though as this is unlikely to be device-dependent.

I did not change that hint, actually I want the main loop to be blocked, the issue is just that it consumes the entire CPU core unless I prevent SDL_PollEvent() from being called when the thread has already been blocked (not sure how it can actually call the function when the thread is blocked though, maybe there is a race condition somewhere?). With my workaround in place the behavior is as expected and CPU utilization is very low when the application is stopped.

I just used the top command from an adb shell to check the CPU usage so nothing fancy.

There are multiple threads in my application but I tried to disable all except the main thread when investigating this issue and the behavior was identical. It really seems like waiting for the semaphore is eating all the CPU cycles for some reason. And likewise, when having all threads enabled and using my workaround CPU usage is close to zero when stopped so the multi-threading in my application is not what is causing the problem.

Please let me know if there is anything specific you want me to test as I can very easily reproduce the problem!

**Comment by 1bsyl** [comment]
Yes please give a try on the emulator. I mean this should be quite straight forward. I wont be able to give try until a few days.

Btw. Maybe you can get the thread name that is eating all the cpu?

If your app is multi threaded. You should get the events will sdl_pollevent and wait for the event:  will_enter_background. When you get it, pause all other threads. And the next call to sdl pollevents will be the blocking one.

(Resume them when you get the event did-enter-foreground)

**Comment by leonstyhre** [comment]
I've now tested on another ARM device as well as the Android emulator (x86) and it's the same problem on both of these. The issue also occurs when locking the screen which is of course also stopping the application.

The thread that is consuming all the CPU cycles is SDLAudioP2, here's the relevant line from top:
`11150 u0_a252       4 -16  14G 207M 100M R 99.0   3.7   0:33.45 SDLAudioP2      org.es_de.frontend`

And thanks for letting me know about the SDL_APP_WILLENTERBACKGROUND event, it seems like I can reliably toggle the flag (an atomic bool) to not process events using this so that I can avoid having to override the onPause() callback. It still doesn't solve the problem though and I have to keep overriding onResume() to be able to get back to polling events after the application is running again.

**Comment by slouken** [comment]
Ryan, can you look at this? Shouldn't we be pausing audio in the background on Android?

**Comment by 1bsyl** [comment]
Audio should be paused.(it used to be)

The default audio back end is aaudio. 
What happens if you use the backebd openSLES?
(Comment out the backend in sdl_audio.c)

**Comment by 1bsyl** [comment]
Pausing here:
https://github.com/libsdl-org/SDL/blob/9c3e831e33062b605f6ee85bb790a3be99bb52e5/src/video/android/SDL_androidevents.c#L112


Backend to comment
https://github.com/libsdl-org/SDL/blob/9c3e831e33062b605f6ee85bb790a3be99bb52e5/src/audio/SDL_audio.c#L57

Aaudio pause. Maybe an issue...
https://github.com/libsdl-org/SDL/blob/9c3e831e33062b605f6ee85bb790a3be99bb52e5/src/audio/aaudio/SDL_aaudio.c#L437

**Comment by icculus** [comment]
> @icculus, can you look at this? Shouldn't we be pausing audio in the background on Android?

As Sylvain said, it won't pause audio until the next call to SDL_PumpEvents, where we will pause playback via AAudio, and then hold the device lock until things return, which should freeze out the audio thread.

Can we confirm that AAUDIO_PauseDevices() is getting called?

**Comment by leonstyhre** [comment]
So I can confirm that switching from AAudio to openslES completely eliminates the problem.
I simply changed the following line in CMakeLists.txt by setting the option to 0 instead of 1:
`set(SDL_AUDIO_DRIVER_AAUDIO 1)`

However there is an equally interesting improvement by changing to openslES and that is that sound latency in my application is radically improved. I'm currently in the process of porting my application to Android and it was immediately obvious when I got it up and running that there is an immense amount of latency when navigating the application, i.e. sound lags behind quite a bit. And although there is still noticeably more latency (with openslES) on Android than on Linux, macOS and Windows it's at least now within reason and good enough for the application to be fully usable.

I'm not familiar with either AAudio or openslES, what is the benefit of using AAudio in the first place, or in other words could I simply keep using openslES to get the benefit of the reduction in latency as well as fixing the 100% CPU utilization problem?

Although it would of course be best if you could improve latency for the AAudio driver so it's equivalent to openslES :)

**Comment by icculus** [comment]
I don't have an opinion of what backend is better, if OpenSLES is widely available and working better, maybe we make that the default and be done with it, but @1bsyl will know _way_ better than I would about this.

But also, there are like a million settings in AAudio and it's possible we aren't setting something optimally, and it's not always clear what the right thing to do is. At least, that was my experience when I was reworking things for SDL3.

**Comment by 1bsyl** [comment]
Aaudio is supposed to be the recent API but opensles works also

For aaudio, there were so recent re factoring
It used to be different.

https://github.com/libsdl-org/SDL/blob/6d9ccbb3c77f42be13933d3ad9fa2dd9219e3894/src/audio/aaudio/SDL_aaudio.c#L338

Now it only calls finddevice and PauseOneDevice().
So is PauseOneDevice() really called? Can you check also sub call to pause function? And also locks?



(Assuming you have the recent version...)

There was also a recent commit about latency. Maybe try to re add low latency

https://github.com/libsdl-org/SDL/commit/99fa0d6caeca00e2968beb364cd0493bdd560e61

**Comment by leonstyhre** [comment]
Thanks for the info! To me it sounds like long term AAudio would be the way to go as I guess it's the more modern option, but due to the issues with latency and CPU utilization I will keep using OpenSL ES for the time being.
If there is anything else you want me to test then please let me know.

Btw, any plans to use the Oboe library in the future instead of AAudio and OpenSL ES directly?

And sorry for going on about the semaphore thing, the issue obviously had nothing to do with that although it really looked like that to me, at first.

**Comment by 1bsyl** [comment]
Not sure about the oboe lib. I dont really remember how it fits in the audio stack. But if it makes sense and someone comes with a backend. It can be added for sure.


Yes. Please can you try with aaudio backend:
Tell which version use you. Sdl2 or latest sdl3?
If sdl3:
- Try to re enable the previous #ifdef to see if this solves latency.
- add traces to see if the pause aaudio functions are really called. (PauseOneDevice() function of aaudio should call AAudioStream_requestPause())
Btw: you are using only 1 device? And no audio capture?

**Comment by 1bsyl** [comment]
> Oboe is a C++ wrapper that provides an API that closely resembles AAudio. It calls AAudio when it is available, and falls back to OpenSL ES if AAudio is not available. AAudio is a new Android C API introduced in the Android O release. It is designed for high-performance audio applications that require low latency.

So it's a lib on top of aaudio or opensles. Not sure we need it!

**Comment by leonstyhre** [comment]
Thanks for the info, so it sounds like Oboe is not really necessary to support then. Anyway, regarding the latency I have tested primarily on 2.28.5 which does not include the latency patch/change mentioned, so I guess that's not the cause of the high latency then.

I'm testing on two separate devices, a Google Pixel 4a and an Ayn Odin 2 and they have the exact same behavior, i.e. very high latency with AAudio as well as the 100% CPU issue, both of which are fixed by reverting to OpenSL ES.

**Comment by 1bsyl** [comment]
You dont need to test with capture. I was just asking if your were using it.(unlikelyl.

Please let me know if the pause request is called.

**Comment by 1bsyl** [comment]
Ok you re on sdl2. So this is the older aaudio code.
Maybe semaphore are badly implemented on android :/
I should have a look.

**Comment by 1bsyl** [comment]
Not sure about the latency.
Maybe some param to tweak.
Maybe you need to add the line
https://github.com/libsdl-org/SDL/blob/e3d50619f89ed747f3e15e8410c13d5dd47956d7/src/audio/aaudio/SDL_aaudio.c#L316

**Comment by leonstyhre** [comment]
I just tried to build 3.0.0 using commit e3d50619f89ed747f3e15e8410c13d5dd47956d7 but unfortunately I got build errors like the following:
`src/video/android/SDL_android_video_capture.c:122:14: error: use of undeclared identifier 'SDL_PIXELFORMAT_RGB888_renamed_SDL_PIXELFORMAT_XRGB8888'`

So instead building from the tag prerelease-3.0.0 I can confirm that the latency is equally high for AAudio as in 2.28.5 and the 100% CPU utilization issue is present as well.

In addition to this the AAudio driver has a very strange problem with 3.0.0 which is that the volume fluctuates wildly. Over a cycle of around 3-5 seconds or so it goes up and down a lot. I'm not sure I've experienced this in 2.28.5 although it's possible. Switching to OpenSL ES fixes this issue a well in addition to the latency and CPU utilization problems.

Another observation on both 2.28.5 and 3.0.0 is that the sound volume is much lower when using OpenSL ES as compared to AAudio but I'm not sure if this is an error, it's still interesting as it's a large difference.

I will revert to 2.28.5 now, I prefer to keep using a stable release until 3.0.0 is ready for production. But I'll be very happy to try out things in 3.0.0 if it can help to resolve the audio problems.

**Comment by slouken** [comment]
The audio code in main (3.0 with no tag) is vastly different from SDL 2.x and the prerelease tag. I think I fixed the compile error you were getting, can you try the main code again?

**Comment by leonstyhre** [comment]
I see, then I'll for sure test it. It works fine to compile the library now but there are a lot of linker errors such as the following:

```
ld: error: undefined symbol: ACameraManager_create
ld: error: undefined symbol: AImageReader_getWindow
ld: error: undefined symbol: ACameraDevice_createCaptureRequest

**Comment by slouken** [comment]
Ah, thanks. I'll have a chance to look at the linker errors next week. I'll let you know when it's building correctly.

**Comment by leonstyhre** [comment]
Ok, great! Happy New Year!

**Comment by icculus** [comment]
This is probably on my plate, since I'm messing with the camera stuff off in a branch.

**Comment by 1bsyl** [comment]
I tried, with an intermediate of SDL3 (close to what SDL2 looks like). version shouldn't really matter.
I didn't see the 100 %cpu with AAudio :(

@leonstyhre 
can you try without playing any audio ?  (I mean just don't feed the audio driver with music)

**Comment by leonstyhre** [comment]
I have not been able to test with a recent SDL3 build due to the linker errors, but with release-2.28.5 and prerelease-3.0.0 the problem is present. I just tested with starting the application on my phone without outputting any audio at all, then immediately pressing the power button to turn off the screen/lock the device. This triggered the 100% CPU utilization by SDLAudioP2.

The most important part of my audio code including the subsystem initialization can be found here if it could help in any way:

https://gitlab.com/es-de/emulationstation-de/-/blob/master/es-core/src/AudioManager.cpp?ref_type=heads

I'm btw probably going to update ES-DE to use SDL_mixer in a future release, but I have no idea if that would change anything when it comes to this problem...

**Comment by 1bsyl** [comment]
I don't really know :/
I think it's ok to test with latest SDL2 and not SDL3. should be the same bug. the sdl3 I have, even if different, is closed to sdl2.

I would really comment out the inside of each function:

mixAudio
play
stop
setupAudioStream
processStream
clearStream

( maybe some SDL_LockAudioDevice() prevent from pausing correctly)

**Comment by leonstyhre** [comment]
I tried to comment out all the code you mentioned and it's still the exact same behavior unfortunately. This is with 2.28.5.

**Comment by 1bsyl** [comment]
ok...
can you try with the most basic example possible ? 
this is SDL2: 
[main.c.txt](https://github.com/libsdl-org/SDL/files/13886802/main.c.txt)
(feel free to reduce it / add something inside so that we can isolate the issue)

to see if, that happens ?

**Comment by leonstyhre** [comment]
I have tested this now, and with this code the issue does not occur, i.e. when putting the screen to sleep or otherwise stopping the activity the CPU utilization goes to ~0% as expected even when using AAudio.

**Comment by 1bsyl** [comment]
@leonstyhre 
I'd say, try to increment this small example to reach a state where you can reproduce the issue.

or, the other way, comment out everything in your game until you make the issue disappear. So we know what is the last step causing the issue.

**Comment by icculus** [comment]
(To be fair, that example code doesn't open an audio device, so there'd be no audio device thread to take all the CPU.  :)  )

**Comment by 1bsyl** [comment]
yes, that's true. @leonstyhre you can add something like this, which should run some audio ?

```
SDL_AudioSpec wanted;
SDL_AudioSpec spec;
SDL_AudioDeviceID devid;

SDL_zero(wanted);
wanted.freq = 44100;
wanted.format = AUDIO_F32SYS;
wanted.channels = 1;
wanted.samples = 4096;
wanted.callback = NULL;

SDL_zero(spec);

// open
devid = SDL_OpenAudioDevice(NULL, SDL_FALSE, &wanted, &spec, SDL_AUDIO_ALLOW_ANY_CHANGE);
if (!devid_out) {
  SDL_Log("Couldn't open an audio device for playback: %s!", SDL_GetError());
  SDL_Quit();
  exit(1);
}

// play
SDL_PauseAudioDevice(devid, SDL_FALSE);

// fill some audio
{
  Uint8 buf[1024];
  SDL_memset(buf, 0, 1024);
  SDL_QueueAudio(devid, buf, 1024);
}

```

**Comment by icculus** [comment]
I have this reproducing in the emulator, I think; looking closer.

**Comment by icculus** [comment]
okay, here's what's happening:

We're correctly calling aaudio_PauseDevices() when we hit the Home button to put the app in the background, but it has this code:

```c
        if (SDL_AtomicGet(&audioDevice->paused)) {
            /* The device is already paused, leave it alone */
            private->resume = SDL_FALSE;
        } else {
            SDL_LockMutex(audioDevice->mixer_lock);
            SDL_AtomicSet(&audioDevice->paused, 1);
            private->resume = SDL_TRUE;
        }
```

We _need_ that SDL_LockMutex() call to block the audio thread, otherwise it apparently just flies at top speed forever, presumably throwing away audio and likely failing to wait on the device before sending more.

But it looks like it considers the "pause" flag as good enough and doesn't lock the device. The problem is that SDL isn't _really_ pausing anything in this case, it just doesn't call the app's audio callback and feeds the audio device silence. So that check is wrong.

OpenSL ES isn't having this problem because its WaitDevice implementation waits for the OS to fire a callback that posts a semaphore, which it _doesn't_ fire because we paused the device, I assume...so the device thread just blocks, waiting for that semaphore.

AAudio doesn't implement WaitDevice (yikes!), so I assume that PlayDevice blocks until it can feed the audio to the device, but in the background I assume this fails, and since we are ignoring the error message (beyond a debug log that is disabled), this just continues as fast as possible.

SDL3 _requires_ a WaitDevice implementation, and it looks like we moved to AAudio's callback interfaces there, _and_ we lock the device too...so it _shouldn't_ be broken there.

**Comment by icculus** [comment]
okay, I've pushed a fix, but SOMEONE PLEASE TEST THIS to make sure it works. I'm just working with an audio device that is playing silence in a crappy test app, so I need to make sure this doesn't break normal usage.

**Comment by 1bsyl** [comment]
I can test next week, but I wasn't able to reproduce before with a real app. I am using SDL_mixer / and occasionnaly playing sound (no endless music loop).

btw, if the fix, is ok, maybe the java audio/android would need also this!
https://github.com/libsdl-org/SDL/blob/SDL2/src/audio/android/SDL_androidaudio.c#L148

**Comment by leonstyhre** [comment]
@icculus Wow that was not an obvious issue! I have tested ES-DE with your patch and I can confirm that it resolves the problem with the 100% CPU utilization.

But as there is still very significant audio latency when using AAudio I'm sticking to OpenSL ES for the time being. It would be really nice if you could investigate that as well though, as I would say that AAudio is almost unusable due to the amount of latency. Do you want me to create a separate issue for that?

**Comment by icculus** [comment]
> but I wasn't able to reproduce before with a real app

Call SDL_PauseAudioDevice(dev, 1) and then put the app in the background.  I reproduced it by dumb luck of not unpausing the device after opening it.

**Comment by icculus** [comment]
> Do you want me to create a separate issue for that?

Yes, please. Also note that it reproduces with ES-DE specifically so we have a known test case to examine!

**Comment by icculus** [comment]
> btw, if the fix, is ok, maybe the java audio/android would need also this!

Oh, yes, that needs the fix too.

**Comment by icculus** [comment]
This is fixed for the other Android backend now, too, in 1e13aa2c98b7740c4747af6651df24e977c97e7f.

I'm closing this as resolved (we'll deal with any AAudio latency in another issue).

**Comment by leonstyhre** [comment]
I have now also tested OpenSL ES with commit 1e13aa2c98b7740c4747af6651df24e977c97e7f and I can confirm that it seems to still work as before (i.e. still working fine).
